### ISSUE TITLE
What does "SOCKS4 tunnel failed, connection closed" mean?

### ISSUE THREAD

**Issue by cowwoc** [eclipse/jetty.project#7822]
**Jetty version**
11.0.8

**Java version**
17

**Question**
While sending out many HTTP requests over a SOCKS proxy I get `SOCKS4 tunnel failed, connection closed`. This is coming from https://github.com/eclipse/jetty.project/blob/cb127793e5d8b5c5730b964392a9a905ba49191d/jetty-client/src/main/java/org/eclipse/jetty/client/Socks4Proxy.java#L164

The stack-trace looks like this:

```
java.io.IOException: SOCKS4 tunnel failed, connection closed
	at org.eclipse.jetty.client.Socks4Proxy$Socks4ProxyConnection.onFillable(Socks4Proxy.java:164)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:319)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:412)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:381)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:268)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.lambda$new$0(AdaptiveExecutionStrategy.java:138)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:407)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:894)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1038)
	... 1 more
```

I don't understand what the error means. Who is closing the connection?

Does it mean that HTTP client is unable to establish a connection to the SOCKS proxy?
Is it able to establish a connection but the proxy drops the connections mid-request?
Is it able to establish a connection to the proxy, but then the target server disconnects on the proxy?

What are the typical scenarios for this error occurring?

Thank you.

**Comment by cowwoc** [comment]
I ran across one scenario where this happens:

* HttpClient is configured to use a SOCKS4 proxy over an SSH connection.
* HttpClient attempts to send a request to a non-existent hostname.
* The operation fails with `java.io.IOException: SOCKS4 tunnel failed, connection closed`.
* If you read the server-side log at `/var/log/secure` you will see: `sshd: error: connect_to [hostname]: unknown host (Name or service not known)`

In other words, the ssh server closed the connection because a tunnel could not be established to a non-existent endpoint.

I hope this helps other users.

**Comment by sbordet** [comment]
The error you posted means that the Jetty `HttpClient` was trying to read from the SOCKS server, but the connection was closed by the server.

If the SOCKS bytes are sent to a SSH tunnel, then Jetty's `HttpClient` is actually connected to a local SSH "server" that runs the tunnel, so it could be that the SSH connection was closed.

I don't think it is possible to distinguish whether it's the SSH socket that was closed, or if the SSH tunnel is open, but the actual SOCKS server closed the communication.
